singer-commons/src/main/java/com/pinterest/singer/loggingaudit/client/AuditEventKafkaSender.java

```diff
@@ -0,0 +1,340 @@
+   *  creating this deque in the LoggingAuditClient.java
+   *
+   *  The AuditEventKafkaSender dequeue from the beginning of the deque, if does not send out event
+   *  successfully, it will enqueue this event to the beginning of the queue.  Note that, if enqueue
+   *  the failed event at the end of the queue, this event could be processed with quite some delay
+   *  and this is not the behavior we want.  That's we use LinkedBlockingDequeue,
+   *  not ArrayBlockingQueue.
+   */
+  private final LinkedBlockingDeque<LoggingAuditEvent> queue;
+
+  /**
+   * KafkaProducer instance to send audit events
+   */
+  private KafkaProducer<byte[], byte[]> kafkaProducer;
+
+  /**
+   * Serialize key and value to byte[]
+   */
+  private TSerializer serializer = new TSerializer();
+
+  /**
+   *  flag to control the start and stop of the executing thread.
+   */
+  private AtomicBoolean cancelled = new AtomicBoolean(false);
+
+  /**
+   *  topic to store the audit events
+   */
+  private String topic;
+
+  /**
+   *  name of this sender instance
+   */
+  private String name;
+
+  /**
+   *  executing thread
+   */
+  private Thread thread;
+
+  /**
+   *  dequeue wait in seconds
+   */
+  private int dequeueWaitInSeconds = 30;
+
+  /**
+   *  when gracefully shutting down the Sender, this field specifies maximum time for main thread
+   *  to wait, in order to let the sender send out audit events left in the queue if there are any.
+   */
+  private int stopGracePeriodInSeconds = 300;
+
+  /**
+   *   when gracefully shutting down the Sender, the calling thread sleeps for some time and let
+   *   the sender to finish its job.   stopGracePeriodInSeconds /  callingThreadSleepInSeconds  is
+   *   the total rounds the calling thread will wait.
+   */
+  private int callingThreadSleepInSeconds = 30;
+
+  /**
+   *  When sender send audit events to Kafka,  it chooses a random partition and if it fails, it
+   *  will choose another GOOD partition, this retry will happen at most numOfPartitionsToTrySending
+   *  times before dropping the event. Note that, this is different from the retry handled by kafka client
+   *  library when sending an event to a certain partition.
+   */
+  private int numOfPartitionsToTrySending = 3;
+
+  /**
+   *  For each partition, track the number of sending failures happened to this partition.
+   */
+  private Map<Integer, AtomicInteger> partitionErrorCounts = new ConcurrentHashMap<>();
+
+  /**
+   * For each event (identified by LoggingAuditHeaders, key of the map), track the partitions that
+   * has been used to send this events. These partitions are aka bad partitions, because it failed
+   * sending out this event. During retry, new partition not belong the bad partitions  is
+   * randomly chosen. When event is send out successfully or dropped, the corresponding entry in
+   * this map is removed.
+   */
+  private Map<LoggingAuditHeaders, Set<Integer>> eventBadPartitionsMap = new ConcurrentHashMap<>();
```

Comment:
As we have only one audit topic, do we need to maintain a [ loggingauditheader --> set<bad partitions>] mapping?  if one partition gets offline,  all  auditing messages that send to that partition will fail.   shall we just have a `Set<Integer>` to track the bad partitions?

Reply:
refactored in the latest update:

(1) maintained badPartitions: if event was not sent to one partition successfully, this partition will be added to badPartitions; if succeed, the partition is removed if it was added before.
badPartitions is for all events.  Every retry will choose a different partition that is not in badPartitions. 

(2) maintained  eventTriedCount: for each event, count number of tries, if reach the threshold 3, drop this event.  Note that: each retry is sent out to a partition that is randomly chosen and not in badPartitions;  also, event will be put into this map eventTriedCount only if it failed, this is done in the callback.